plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseTableProperties.java

```diff
@@ -36,26 +42,120 @@
+    public static final String ORDER_BY_PROPERTY = "order_by"; //required
+    public static final String PARTITION_BY_PROPERTY = "partition_by"; //optional
+    public static final String PRIMARY_KEY_PROPERTY = "primary_key"; //optional
+    public static final String SAMPLE_BY_PROPERTY = "sample_by"; //optional
+
+    public static final ClickHouseEngineType DEFAULT_TABLE_ENGINE = ClickHouseEngineType.LOG;
 
     private final List<PropertyMetadata<?>> tableProperties;
 
     @Inject
     public ClickHouseTableProperties()
     {
         tableProperties = ImmutableList.of(
-                stringProperty(
+                enumProperty(
                         ENGINE_PROPERTY,
                         "ClickHouse Table Engine, defaults to Log",
+                        ClickHouseEngineType.class,
                         DEFAULT_TABLE_ENGINE,
+                        false),
+                new PropertyMetadata<>(
+                        ORDER_BY_PROPERTY,
+                        "columns to be the sorting key. Required",
+                        new ArrayType(VARCHAR),
+                        List.class,
+                        ImmutableList.of(),
+                        false,
+                        value -> ImmutableList.copyOf(((Collection<?>) value).stream()
+                                .map(name -> ((String) name).toLowerCase(ENGLISH))
+                                .collect(Collectors.toList())),
+                        value -> value),
+                new PropertyMetadata<>(
+                        PARTITION_BY_PROPERTY,
+                        "columns to be the partition key. Optional",
+                        new ArrayType(VARCHAR),
+                        List.class,
+                        ImmutableList.of(),
+                        false,
+                        value -> ImmutableList.copyOf(((Collection<?>) value)),
+                        value -> value),
+                new PropertyMetadata<>(
+                        PRIMARY_KEY_PROPERTY,
+                        "columns to be the primary key. Optional",
+                        new ArrayType(VARCHAR),
+                        List.class,
+                        ImmutableList.of(),
+                        false,
+                        value -> ImmutableList.copyOf(((Collection<?>) value).stream()
+                                .map(name -> ((String) name).toLowerCase(ENGLISH))
+                                .collect(Collectors.toList())),
+                        value -> value),
+                stringProperty(
+                        SAMPLE_BY_PROPERTY,
+                        "An expression for sampling. Optional.",
+                        null,
                         false));
     }
 
-    public static Optional<String> getEngine(Map<String, Object> tableProperties)
+    public static ClickHouseEngineType getEngine(Map<String, Object> tableProperties)
+    {
+        requireNonNull(tableProperties);
```

Comment:
In case when `tableProperties` is null, this line produces `NullPointerException` with no message, just as the next line would do, so it's OK to skip it.

If you want to do validation, please add a message like
```
requireNonNull(tableProperties, "tableProperties is null");
```

you can use intellij template do write this conveniently, see https://gist.github.com/findepi/b597d2bc91e5a30c2a44fd65fe252f0b

Reply:
That's cool!